File: src/level_info.rb
class LevelInfo
  attr_accessor :filename, :title, :ini_file
  
  def initialize filename
    @filename = filename
    @ini_file = INIFile.new(filename)
    @title       = t(@ini_file['Info', 'Title']  || 'Unbenannte Karte')
    @difficulty  = t(@ini_file['Info', 'Skill']  || '')
    @description = t(@ini_file['Info', 'Desc']   || '')
    @author      =  (@ini_file['Info', 'Author'] || '')
    
    stars_goal = (@ini_file['Map', 'StarsGoal'] || 100).to_i
    if stars_goal == 0 then
      @goal = t "Durchkommen"
    else
      @goal = "#{stars_goal} #{t 'Sterne einsammeln'}"
    end
    
    hostages = []
    obj = 0
    while obj_desc = @ini_file['Objects', obj] do
      if obj_desc[0, 2] == ('%02X' % ID_CAROLIN) then
        obj_desc_extended = @ini_file['Objects', "#{obj}Y"] || '|'
        hostages << (obj_desc_extended.split('|').last || 'Carolin')
      end
      obj += 1
    end
    if hostages.size == 1 then
      @goal += " #{t 'und'} #{hostages[0]} #{t 'retten'}"
    elsif hostages.size > 1 then
      @goal += " #{t 'und'} #{hostages.size} #{t 'Gefangene retten'}"
    end
    
    # TODO highscore = ...
  end
  
  def draw y, active
    if active then
      draw_rect 0, y + 1, 631, 98, 0xff603000
    end
    
    draw_rect 0, y, 631, 1, 0x003000
    # TODO (Highscore) / (not beaten yet)
    draw_string @title, 5, y + 7, 255
    draw_right_aligned_string @difficulty, 626, y + 7, 255
    draw_string @description, 5, y + 30, 192
    draw_string @goal, 5, y + 53, 128
    draw_string @author, 5, y + 76, 80
    draw_rect 0, y + 99, 631, 1, 0x006000
  end
  
  FIRST_LEVEL = 'jr_Gemuetlicher_Aufstieg.pml'
  
  def <=> other
    if File.basename(self.filename) == FIRST_LEVEL then
      -1
    elsif File.basename(other.filename) == FIRST_LEVEL then
      +1
    else
      title <=> other.title
    end
  end
end

File: src/LevelInfo.cpp
#include "LevelInfo.hpp"
#include "Constants.hpp"
#include "Options.hpp"
#include "helpers/Graphics.hpp"
#include "helpers/I18n.hpp"
#include "helpers/String.hpp"
#include <algorithm>
#include <filesystem>
#include <fstream>
#include <string_view>

LevelInfo::LevelInfo(const std::string& filename)
    : filename(filename),
      ini_file(std::ifstream(filename))
{
    title = ini_file.string("Info", "Title").value_or("Unbenanntes Level");
    difficulty = ini_file.string("Info", "Skill").value_or("");
    description = ini_file.string("Info", "Desc").value_or("");
    author = ini_file.string("Info", "Author").value_or("");

    // The goal is composed from translated() parts, like t() in the Ruby version. The English
    // parts are worded so that they work in the same order as the German original.
    goal = ini_file.string("Map", "StarsGoal").value_or("100");
    if (goal == "0") {
        goal = t("Durchkommen");
    }
    else {
        goal += " " + t("Sterne einsammeln");
    }

    int hostages_count = 0;
    std::optional<std::string> hostage_name;
    for (int i = 0;; ++i) {
        const std::optional obj = ini_file.string("Objects", std::to_string(i));
        if (!obj) {
            break;
        }
        if (obj->starts_with(byte_to_hex(ID_HOSTAGE))) {
            hostages_count += 1;
            const std::optional extra_data = ini_file.string("Objects", std::to_string(i) + "Y");
            if (extra_data && extra_data->length() > 2) {
                hostage_name = extra_data->substr(2);
            }
        }
    }

    if (hostages_count == 1) {
        goal += " " + t("und") + " " + hostage_name.value_or("Carolin");
        if (const std::string rescue = t("retten"); !rescue.empty()) {
            goal += " " + rescue; // "retten" trails the name in German, but is empty in English
        }
    }
    else if (hostages_count > 1) {
        goal += " " + t("und") + " " + std::to_string(hostages_count) + " "
            + t("Gefangene retten");
    }

    hiscore = load_hiscore(filename);
}

std::vector<LevelInfo> LevelInfo::list_levels()
{
    std::vector<LevelInfo> levels;

    for (const auto& entry : std::filesystem::directory_iterator("levels")) {
        if (entry.path().extension() == ".pml") {
            levels.push_back(LevelInfo(entry.path().string()));
        }
    }

    static constexpr std::string_view first_level = "jr_Gemuetlicher_Aufstieg.pml";
    std::ranges::sort(levels, std::less(), [](const LevelInfo& info) {
        return std::tuple { !info.filename.ends_with(first_level), info.difficulty };
    });

    return levels;
}

#include <doctest.h>

TEST_CASE("LevelInfo")
{
    const std::vector<LevelInfo> levels = LevelInfo::list_levels();
    CHECK(levels.size() >= 10);

    SUBCASE("Gemütlicher Aufstieg comes first")
    {
        // Check if the first level is indeed the one specified in the code
        static constexpr std::string_view first_level = "jr_Gemuetlicher_Aufstieg.pml";
        CHECK(levels.at(0).filename.ends_with(first_level));
        CHECK(levels.at(0).title == "Gemütlicher Aufstieg");
        CHECK(levels.at(0).difficulty == "Sehr einfach");
        CHECK(levels.at(0).author == "Julian Raschke, julian@raschke.de, www.petermorphose.de");
    }

    SUBCASE("Gravialistan can be found")
    {
        // Check another level
        const auto it = std::ranges::find_if(levels, [](const LevelInfo& info) {
            return info.filename.ends_with("sl_Gravialistan.pml");
        });
        CHECK(it != levels.end());
        if (it != levels.end()) {
            CHECK(it->title == "Gravialistan");
            // The goal is language-dependent; accept the German and the English composition.
            CHECK((it->goal == "100 Sterne einsammeln und Tanja retten"
                   || it->goal == "100 stars to collect & need to rescue Tanja"));
        }
    }
}
